WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift

```
@@ -205,4 +258,89 @@ private extension ProductFormRemoteActionUseCase {
         }
         stores.dispatch(passwordUpdateAction)
     }
+
+    func duplicateVariations(_ variationIDs: [Int64],
+                             from oldProductID: Int64,
+                             to newProduct: EditableProductModel,
+                             onCompletion: @escaping (Result<EditableProductModel, ProductUpdateError>) -> Void) {
+        Task { [weak self] in
+            guard let self = self else { return }
+            for id in variationIDs {
+                guard let variation = await self.retrieveProductVariation(id: id, siteID: newProduct.siteID, productID: oldProductID) else {
+                    continue
+                }
+                let newVariation = CreateProductVariation(regularPrice: variation.regularPrice ?? "", attributes: variation.attributes)
+                await self.duplicateProductVariation(newVariation, parent: newProduct)
+            }
+
+            let updatedProduct: EditableProductModel = await {
+                guard let productModel = await retrieveProduct(id: newProduct.productID, siteID: newProduct.siteID) else {
+                    return newProduct
+                }
+                return EditableProductModel(product: productModel)
+            }()
+
+            await MainActor.run {
+                onCompletion(.success(updatedProduct))
+            }
+        }
+    }
+
+    func retrieveProduct(id: Int64, siteID: Int64) async -> Product? {
```

❓ I am curious about returning optional here. Do you think throwing instead of returning an optional complicates the code at the calling place?

---

The product is already duplicated before we proceed to duplicate variations, so I'm not sure whether to make this fail if there is a problem in retrieving variations or the updated product. 

But this retrieval is needed to update the UI with the updated product, so maybe it makes sense to throw an error when the data is not updated.